src/controller/SetUpCodeParser.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implementation of SetUp Code Parser, a class
+ *      that implements parsing manual and QRCode for connecting
+ *      to a CHIP Device.
+ *
+ */
+
+#include <controller/SetUpCodeParser.h>
+
+#include <controller/CHIPDeviceController.h>
+#include <lib/mdns/Resolver.h>
+#include <lib/support/CodeUtils.h>
+
+namespace chip {
+namespace Controller {
+
+CHIP_ERROR SetUpCodeParser::PairDevice(NodeId remoteId, const char * setUpCode)
```

Comment:
I think you are saying that we should leave this pairing concept as it currently exists in the sdk in place, where the sdk's pairing is approximately:

# discover
# commission
# persist paired device in kvstore

Yes, this is something above / beyond / out-of-scope from the spec perspective.  It would be good to understand that distinction and move the sdk to reflect that e.g. with an appropriate division between code in ./examples and ./src.

Reply:
Agree things could be separated, however - the functionality here is definitely something of value in the SDK to some members, and nobody is required to use it - if some adopters don't require this functionality.  Agree with @cecille that we could separate pairing from commissioning, however - that also could be separate. 

However - a simple API to commission a device from a setup code, and a nodeID seems like a good, which I think this is doing.

Reply:
Closing the conversation as per @cecille comment. For my part I do not have any strong opinion. Naming is hard...